book-content/chapters/10-deriving-types.md

```diff
@@ -0,0 +1,1460 @@
+] as const;
+
+// ---cut---
+type AllPrograms = (typeof programModes)[keyof typeof programModes];
+//   ^?
+```
+
+The additional stuff being extracted causes the test to fail because it is only expecting the original values instead of numbers and functions.
+
+Recall that we can access the first element using `programModes[0]`, the second element using `programModes[1]`, and so on. This means that we could use a union of all possible index values to extract the values from the `programModes` array:
+
+```typescript
+type AllPrograms = (typeof programModes)[0 | 1 | 2 | 3 | 4 | 5];
+```
+
+This solution makes the test pass, but it doesn't scale well. If the `programModes` array were to change, we would need to update the `AllPrograms` type manually.
+
+Instead, we can use the `number` type as the argument to the indexed access type to represent all possible index values:
+
+```typescript
+type AllPrograms = (typeof programModes)[number];
+```
+
+Now new items can be added to the `programModes` array without needing to update the `AllPrograms` type manually. This solution makes the test pass as expected, and is a great pattern to apply in your own projects.
+
+## Deriving Types From Functions
+
+So far, we've only looked at deriving types from objects and arrays. But deriving types from functions can help solve some common problems in TypeScript.
+
+### `Parameters`
+
+The `Parameters` utility type extracts the parameters from a given function type and returns them as a tuple.
+
+For example, this `sellAlbum` function takes in an `Album`, a `price`, and a `quantity`, then returns a number representing the total price:
+
+```typescript
+function sellAlbum(album: Album, price: number, quantity: number) {
+  return price * quantity;
+}
+```
+
+Using the `Parameters` utility type, we can extract the parameters from the `sellAlbum` function and assign them to a new type:
+
+```ts twoslash
+type Album = {
+  title: string;
+  artist: string;
+  releaseYear: number;
+};
+
+function sellAlbum(album: Album, price: number, quantity: number) {
+  return price * quantity;
+}
+// ---cut---
+type SellAlbumParams = Parameters<typeof sellAlbum>;
+//   ^?
+```
+
+Note that we need to use `typeof` to create a type from the `sellAlbum` function. Passing `sellAlbum` directly to `Parameters` won't work on its own because `sellAlbum` is a value instead of a type:
+
+```ts twoslash
+// @errors: 2749
+type Album = {
+  title: string;
+  artist: string;
+  releaseYear: number;
+};
+function sellAlbum(album: Album, price: number, quantity: number) {
+  return price * quantity;
+}
+// ---cut---
+type SellAlbumParams = Parameters<sellAlbum>;
+```
+
+This `SellAlbumParams` type is a tuple type that holds the `Album`, `price`, and `quantity` parameters from the `sellAlbum` function.
+
+If we need to access a specific parameter from the `SellAlbumParams` type, we can use indexed access types:
+
+```typescript
+type Price = SellAlbumParams[1]; // number
+```
+
```

Comment:
an extra example here could also who how useful it might be when building layers of functions:
```ts
function inner(a: string, b: number) {}

function outer(extra: boolean, ...args: Parameters<typeof inner>) {}

outer;
// ^?
```